packages-next/fields-document/ARCHITECTURE.md

```diff
@@ -0,0 +1,53 @@
+# Document editor architecture
+
+The document editor is using [Slate](https://github.com/ianstormtaylor/slate). To learn about Slate generally, go through https://docs.slatejs.org/walkthroughs and https://docs.slatejs.org/concepts. The sections on [Nodes](https://docs.slatejs.org/concepts/02-nodes) and [Locations](https://docs.slatejs.org/concepts/03-locations) contain especially important terminology.
+
+## Plugins
+
+For normalization and handling user input and some other things(you can see what they are by looking at the type definition for the `Editor` in Slate), you write [Slate "plugins"](https://docs.slatejs.org/concepts/07-plugins), plugins are not some special API, it means accepting an `Editor` object and returning it but modifying some of the properties.
+
+## Rendering
+
+The rendering of the editor is mostly like any other bit of React code. See https://docs.slatejs.org/concepts/08-rendering for how rendering the document works in Slate.
+
+The editor is very performance-sensitive because it will re-render on every keystroke so keep these things in mind while writing code for rendering the editor:
+
+- Renderers for elements/leaves should never use `useSlate` or `useToolbarState` because they will change on every editor change, if they need to get the editor to do something in response to a user action, they should use `useStaticEditor` which will not cause re-renders
```

Comment:
```suggestion
- Renderers for elements/leaves should never use `useSlate` or `useToolbarState` because they will change on every editor change. If they need to get the editor to do something in response to a user action, they should use `useStaticEditor` which will not cause re-renders.
```